### PR TITLE
EIP140 - Implement REVERT opcode

### PR DESCRIPTION
**Pull request by gumb0**
https://github.com/ethereum/cpp-ethereum/issues/3618
https://github.com/ethereum/EIPs/pull/206

### CODE REVIEW

File: libevm/VM.cpp [hunk 1]
```diff
@@ -233,6 +233,24 @@ void VM::interpretCases()
...
+				throwBadInstruction();
+
+			m_newMemSize = memNeed(*m_SP, *(m_SP - 1));
+			updateMem();
+			ON_OP();
+			updateIOGas();
+
+			uint64_t b = (uint64_t)*m_SP--;
+			uint64_t s = (uint64_t)*m_SP--;
+			owning_bytes_ref output{ move(m_mem), b, s };
```
Comment (chriseth) [inline libevm/VM.cpp]: `return` above just assigns to `m_output` - why did you choose to put it into the exception?

Comment (gumb0) [inline libevm/VM.cpp]: Because `m_output` is returned to Executive through return value of `VM::exec()`

And we need a way to both return output data and to tell Executive to revert the state.

We could either return pair/tuple from `VM::exec()` or throw exception with data.

Comment (gumb0) [inline libevm/VM.cpp]: Exception is not ideal, since it's not really an exceptional situation, just another valid opcode.
But I think the code is cleaner this way (and after all REVERT is conceptually described as 'cheap throw')

File: libevmcore/Instruction.h [hunk 2]
```diff
@@ -188,6 +188,7 @@ enum class Instruction: uint8_t
 	CALLCODE,           ///< message-call with another account's code only
 	RETURN,             ///< halt execution returning output data
 	DELEGATECALL,       ///< like CALLCODE but keeps caller's value and sender
+	REVERT = 0xfd,      ///< stop execution and revert state changes, without consuming all provided gas
```
Comment (gumb0) [inline libevmcore/Instruction.h]: Forgot to fill `c_instructions`, `c_instructionInfo` in Instruction.cpp ...

File: libevm/VM.cpp (hunk 1 above)
Comment (chfast) [inline libevm/VM.cpp]: I think we should use `pair<bool, bytes>` (extended `std::optional`), as @chriseth did in precompiled already. We can name this time, and add convenient methods similar to `std::optional`. I think nothing from boost/std directly matches our case here.

If this is a lot refactoring, it can be done in a later PR. Ideally, implement the return type change first, then EIP140.

Comment (gcolvin) [inline libevm/VM.cpp]: I've got a refactoring underway at https://github.com/ethereum/cpp-ethereum/tree/downward-dog, but it doesn't look like it will conflict with this one.

Comment (gcolvin) [inline libevm/VM.cpp]: Talk of a bigger refactoring reminds me that I've never liked the interface to the VM - the way it's a function pretending to be an object, the way that function takes its ration of gas by reference...  I attempted to eliminate the _io_gas reference once and gave up, as it had insinuated itself too deep in the code.

Comment (chfast) [inline libevm/VM.cpp]: Yep. I agree with that. The amount of gas left should be return as well. I'm trying to address that in https://github.com/ethereum/evmjit/blob/develop/include/evm.h#L70-L106, but EVM-C seems to be always postponed.

Comment (gumb0) [inline libevm/VM.cpp]: Ok, I'll do it here. (I was not aware of those precompiled changes)

**Comment by gumb0** [comment]
EVMJIT build is failing, forgot to check it locally...

**Comment by chfast** [comment]
If you need help with it let me know.

File: test/tools/libtestutils/Common.cpp [hunk 3]
```diff
@@ -44,7 +44,7 @@ std::string dev::test::getTestPath()
 	if (ptestPath == NULL)
 	{
 		ctest << " could not find environment variable ETHEREUM_TEST_PATH \n";
-		testPath = "../../../tests";
+		testPath = "../../test/jsontests";
```
Comment (gumb0) [inline test/tools/libtestutils/Common.cpp]: This path won't work on Windows 😞 (there's one subdir more for executables)

Comment (winsvega) [inline test/tools/libtestutils/Common.cpp]: you could use --testpath

Comment (gumb0) [inline test/tools/libtestutils/Common.cpp]: We need to change Appveyor's script then?

Comment (winsvega) [inline test/tools/libtestutils/Common.cpp]: yes

**Comment by gumb0** [comment]
Possible checks/edge cases to pay attention when testing this:
- check that state changes are reverted
- check revert in contract creation code
- check that sender nonce is incremented even after revert
- check that sender nonce is incremented even after revert inside CREATE call
- check that no instructions after REVERT are executed
- check that output data is really returned

File: .gitmodules [hunk 4]
```diff
@@ -5,3 +5,4 @@
 [submodule "test/tests"]
 	path = test/jsontests
 	url = https://github.com/ethereum/tests.git
+	branch = revertTests
```
Comment (winsvega) [inline .gitmodules]: lets not merge before the test coverage

Comment (chfast) [inline .gitmodules]: This is not right. "branch" is only convenient feature that allows you to do `git submodule update --remote`, nothing more. We definitely don't want this to be merged into our default branch.

File: libevm/VM.cpp [hunk 5]
```diff
@@ -233,6 +233,25 @@ void VM::interpretCases()
...
+				throwBadInstruction();
+
+			m_newMemSize = memNeed(*m_SP, *(m_SP - 1));
+			updateMem();
+			ON_OP();
+			updateIOGas();
+
+			uint64_t b = (uint64_t)*m_SP--;
+			uint64_t s = (uint64_t)*m_SP--;
+			m_output = owning_bytes_ref{ std::move(m_mem), b, s };
```
Comment (chfast) [inline libevm/VM.cpp]: Style: if you use `{}` to construct an object use it the same way as `()`.

File: libevm/VMFace.h [hunk 6]
```diff
@@ -33,6 +33,13 @@ ETH_SIMPLE_EXCEPTION_VM(OutOfGas);
 ETH_SIMPLE_EXCEPTION_VM(OutOfStack);
 ETH_SIMPLE_EXCEPTION_VM(StackUnderflow);
 
+struct VmExecResult
+{
+	VmExecResult(owning_bytes_ref&& _output, bool _revertNeeded) : output(std::move(_output)), revertNeeded(_revertNeeded) {}
+
+	owning_bytes_ref output;
+	bool revertNeeded = false;
```
Comment (chfast) [inline libevm/VMFace.h]: This just a comment, but the name for this flag is not right. A VM user does not care about reverts, but cares if the call succeeded or failed. `status` would be better name. Also compare with similar returned type in precompileds.

Comment (gumb0) [inline libevm/VMFace.h]: Hm, but then we'll have two ways to signal `VM::exec()` failure - through exceptions and success return flag - which is not good at all.

Are we planning to get rid of throwing exceptions in favor of returned result then?

Comment (chfast) [inline libevm/VMFace.h]: That would be nice I think. I'm sure @gcolvin agrees :).
This is classic case "should a parser throw exception if syntax is invalid"? IMHO, not, because it is nothing exceptional about detecting syntex errors in parser. It is its _normal_ job. But that's more philosophical...

Anyway. Throw an exception in this PR. Refactoring of this should be separated.

Comment (gumb0) [inline libevm/VMFace.h]: That was the first version of this PR 😃 Ok, I'll try to revive those commits

**Comment by gumb0** [comment]
I revived the version when REVERT results in the special exception (containing output data) being thrown from VM to Executive and made the CALL ended with REVERT be considered failed (pushing 0 to caller's stack).

File: libevm/VMFace.h [hunk 7]
```diff
@@ -25,14 +25,30 @@ namespace dev
 namespace eth
 {
 
-struct VMException: virtual Exception {};
-#define ETH_SIMPLE_EXCEPTION_VM(X) struct X: virtual VMException { const char* what() const noexcept override { return #X; } }
+struct VMException: Exception {};
```
Comment (gumb0) [inline libevm/VMFace.h]: Any reason why these were virtually inherited? It's not like they are going to be inherited several times here in libevm and `VMException` and children don't have data members anyway!

I'm changing it because gcc was issuing a warning for `RevertInstruction` move assignment, which is unsafe with virtual base (virtual base can be move-assigned several times if inherited several times)

**Comment by gumb0** [comment]
Rebased on develop

File: libevm/VMFace.h (hunk 7 above)
Comment (chfast) [inline libevm/VMFace.h]: This is Boost suggestion: http://www.boost.org/doc/libs/1_63_0/libs/exception/doc/using_virtual_inheritance_in_exception_types.html

**Comment by chfast** [comment]
Should I take it from here to implement that in EVMJIT?
